File: Python/13-Heap(PriorityQueue)/Problems/3-Median-of-TwoSortedArrays/Median-of-TwoSortedArrays.py
import heapq
import statistics
import bisect

def find_median_naive(nums1, nums2):
    """
    Approach 1: Naive - Merge and sort.

    Time Complexity: O(m + n log(m + n))
    Space Complexity: O(m + n)
    """
    merged_array = nums1 + nums2
    merged_array.sort()  # Sort the merged array
    n = len(merged_array)
    if n % 2 == 0:
        # Even number of elements, average the two middle elements
        median = (merged_array[n // 2 - 1] + merged_array[n // 2]) / 2.0
    else:
        # Odd number of elements, take the middle element
        median = merged_array[n // 2]
    return median

def find_median_merge(nums1, nums2):
    """
    Approach 2: Merge and find median (optimized merge).

    Time Complexity: O(m + n)
    Space Complexity: O(m + n)
    """
    merged_array = []
    i, j = 0, 0
    while i < len(nums1) and j < len(nums2):
        if nums1[i] <= nums2[j]:
            merged_array.append(nums1[i])
            i += 1
        else:
            merged_array.append(nums2[j])
            j += 1
    # Add the remaining elements of nums1, if any
    merged_array.extend(nums1[i:])
    # Add the remaining elements of nums2, if any
    merged_array.extend(nums2[j:])

    n = len(merged_array)
    if n % 2 == 0:
        median = (merged_array[n // 2 - 1] + merged_array[n // 2]) / 2.0
    else:
        median = merged_array[n // 2]
    return median

def find_median_heap(nums1, nums2):
    """
    Approach 3: Using Heaps (Priority Queues).

    Time Complexity: O((m+n)log(m+n)) - due to heapify, although adding elements is log(k) where k is size of heap.
    Space Complexity: O(m + n)
    """
    merged_array = nums1 + nums2
    heapq.heapify(merged_array) # creates a min heap
    n = len(merged_array)
    for _ in range((n // 2)): # remove the smallest n//2 elements
        heapq.heappop(merged_array)

    if n % 2 == 0:
        return (heapq.heappop(merged_array) + heapq.heappop(merged_array))/2
    else:
        return heapq.heappop(merged_array)

def find_median_pythonic(nums1, nums2):
    """
    Approach 4: Using Python's statistics.median.

    Time Complexity: O(m + n) in the average case,  O(n log n) worst case (due to Timsort)
    Space Complexity: O(1) auxiliary space.
    """
    return statistics.median(nums1 + nums2)

def find_median_binary_search(nums1, nums2):
    """
    Approach 5: Binary Search (Optimal).

    Time Complexity: O(log(min(m, n)))
    Space Complexity: O(1)
    """
    if len(nums1) > len(nums2):
        nums1, nums2 = nums2, nums1  # Ensure nums1 is the shorter array for efficiency

    m, n = len(nums1), len(nums2)
    low, high = 0, m

    while low <= high:
        partition_x = (low + high) // 2
        partition_y = (m + n + 1) // 2 - partition_x

        max_left_x = float('-inf') if partition_x == 0 else nums1[partition_x - 1]
        min_right_x = float('inf') if partition_x == m else nums1[partition_x]
        max_left_y = float('-inf') if partition_y == 0 else nums2[partition_y - 1]
        min_right_y = float('inf') if partition_y == n else nums2[partition_y]

        if max_left_x <= min_right_y and max_left_y <= min_right_x:
            # Found the correct partition
            if (m + n) % 2 == 0:
                return (max(max_left_x, max_left_y) + min(min_right_x, min_right_y)) / 2.0
            else:
                return max(max_left_x, max_left_y)
        elif max_left_x > min_right_y:
            # Move towards the left in nums1
            high = partition_x - 1
        else:
            # Move towards the right in nums1
            low = partition_x + 1
    raise ValueError("Input arrays are not sorted.") # added error

# Example Usage and Output
nums1 = [1, 3]
nums2 = [2, 4]

print("Array 1:", nums1)
print("Array 2:", nums2)

print("\nApproach 1: Naive - Merge and Sort")
print("Median:", find_median_naive(nums1, nums2))

print("\nApproach 2: Merge and Find Median")
print("Median:", find_median_merge(nums1, nums2))

print("\nApproach 3: Using Heaps")
print("Median:", find_median_heap(nums1, nums2))

print("\nApproach 4: Using Python's statistics.median")
print("Median:", find_median_pythonic(nums1, nums2))

print("\nApproach 5: Binary Search")
print("Median:", find_median_binary_search(nums1, nums2))

nums3 = [1, 2]
nums4 = [3, 4, 5]
print("\nArray 3:", nums3)
print("Array 4:", nums4)

print("\nApproach 1: Naive - Merge and Sort")
print("Median:", find_median_naive(nums3, nums4))

print("\nApproach 2: Merge and Find Median")
print("Median:", find_median_merge(nums3, nums4))

print("\nApproach 3: Using Heaps")
print("Median:", find_median_heap(nums3, nums4))

print("\nApproach 4: Using Python's statistics.median")
print("Median:", find_median_pythonic(nums3, nums4))

print("\nApproach 5: Binary Search")
print("Median:", find_median_binary_search(nums3, nums4))

File: JavaScript/13-Heap(PriorityQueue)/Problems/3-Median-of-TwoSortedArrays/Median-of-TwoSortedArrays.js
/*
    Problem: Median of Two Sorted Arrays

    Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.

    The overall run time complexity should be O(log (m+n)).

    Example 1:
    Input: nums1 = [1,3], nums2 = [2]
    Output: 2.00000
    Explanation: merged array = [1,2,3] and median is 2.

    Example 2:
    Input: nums1 = [1,2], nums2 = [3,4]
    Output: 2.50000
    Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.

    Constraints:
    nums1.length == m
    nums2.length == n
    0 <= m <= 1000
    0 <= n <= 1000
    1 <= m + n <= 2000
    -10^6 <= nums1[i], nums2[i] <= 10^6
*/

/**
 * Approach 1: Merge and Find Median
 * Time Complexity: O(m+n)
 * Space Complexity: O(m+n)
 */
function findMedianSortedArrays_merge(nums1, nums2) {
    const merged = [];
    let i = 0;
    let j = 0;

    while (i < nums1.length && j < nums2.length) {
        if (nums1[i] <= nums2[j]) {
            merged.push(nums1[i]);
            i++;
        } else {
            merged.push(nums2[j]);
            j++;
        }
    }

    while (i < nums1.length) {
        merged.push(nums1[i]);
        i++;
    }

    while (j < nums2.length) {
        merged.push(nums2[j]);
        j++;
    }

    const mid = Math.floor(merged.length / 2);
    if (merged.length % 2 === 0) {
        return (merged[mid - 1] + merged[mid]) / 2;
    } else {
        return merged[mid];
    }
}

/**
 * Approach 2: Optimized Merge (Without Extra Array)
 * Time Complexity: O(m+n)
 * Space Complexity: O(1)
 */
function findMedianSortedArrays_optimizedMerge(nums1, nums2) {
    const m = nums1.length;
    const n = nums2.length;
    const total = m + n;
    const mid = Math.floor(total / 2);
    let i = 0;
    let j = 0;
    let left = 0;
    let right = 0;

    for (let k = 0; k <= mid; k++) {
        if (i < m && j < n) {
            if (nums1[i] <= nums2[j]) {
                left = right;
                right = nums1[i];
                i++;
            } else {
                left = right;
                right = nums2[j];
                j++;
            }
        } else if (i < m) {
            left = right;
            right = nums1[i];
            i++;
        } else {
            left = right;
            right = nums2[j];
            j++;
        }
    }

    if (total % 2 === 0) {
        return (left + right) / 2;
    } else {
        return right;
    }
}


/**
 * Approach 3: Binary Search
 * Time Complexity: O(log(min(m,n)))
 * Space Complexity: O(1)
 */
function findMedianSortedArrays_binarySearch(nums1, nums2) {
    const m = nums1.length;
    const n = nums2.length;
    if (m > n) {
        return findMedianSortedArrays_binarySearch(nums2, nums1); // Ensure nums1 is shorter
    }

    let low = 0;
    let high = m;
    const total = m + n;
    const half = Math.floor(total / 2);

    while (low <= high) {
        const partitionX = Math.floor((low + high) / 2);
        const partitionY = half - partitionX;

        const maxLeftX = partitionX === 0 ? -Infinity : nums1[partitionX - 1];
        const minRightX = partitionX === m ? Infinity : nums1[partitionX];

        const maxLeftY = partitionY === 0 ? -Infinity : nums2[partitionY - 1];
        const minRightY = partitionY === n ? Infinity : nums2[partitionY];

        if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
            if (total % 2 === 0) {
                return (Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2;
            } else {
                return Math.max(maxLeftX, maxLeftY);
            }
        } else if (maxLeftX > minRightY) {
            high = partitionX - 1;
        } else {
            low = partitionX + 1;
        }
    }
    throw new Error("Arrays are not sorted.");
}

/**
 * Approach 4: Using a Max Heap and a Min Heap
 * Time Complexity: O((m+n)log(m+n)) - due to heap insertions
 * Space Complexity: O(m+n)
 */
class MaxHeap {
    constructor() {
        this.heap = [];
    }

    insert(value) {
        this.heap.push(value);
        this.bubbleUp(this.heap.length - 1);
    }

    bubbleUp(index) {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.heap[index] <= this.heap[parentIndex]) break;
            this.swap(index, parentIndex);
            index = parentIndex;
        }
    }

    extractMax() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();
        const max = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown(0);
        return max;
    }

    bubbleDown(index) {
        while (true) {
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;
            let largest = index;

            if (leftChild < this.heap.length && this.heap[leftChild] > this.heap[largest]) {
                largest = leftChild;
            }
            if (rightChild < this.heap.length && this.heap[rightChild] > this.heap[largest]) {
                largest = rightChild;
            }
            if (largest === index) break;
            this.swap(index, largest);
            index = largest;
        }
    }

    peek() {
        return this.heap[0];
    }

    size() {
        return this.heap.length;
    }

    swap(i, j) {
        [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
    }
}

class MinHeap {
    constructor() {
        this.heap = [];
    }

    insert(value) {
        this.heap.push(value);
        this.bubbleUp(this.heap.length - 1);
    }

    bubbleUp(index) {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.heap[index] >= this.heap[parentIndex]) break;
            this.swap(index, parentIndex);
            index = parentIndex;
        }
    }

    extractMin() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();
        const min = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown(0);
        return min;
    }

    bubbleDown(index) {
        while (true) {
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;
            let smallest = index;

            if (leftChild < this.heap.length && this.heap[leftChild] < this.heap[smallest]) {
                smallest = leftChild;
            }
            if (rightChild < this.heap.length && this.heap[rightChild] < this.heap[smallest]) {
                smallest = rightChild;
            }
            if (smallest === index) break;
            this.swap(index, smallest);
            index = smallest;
        }
    }

    peek() {
        return this.heap[0];
    }

    size() {
        return this.heap.length;
    }

    swap(i, j) {
        [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
    }
}

function findMedianSortedArrays_heaps(nums1, nums2) {
    const maxHeap = new MaxHeap();
    const minHeap = new MinHeap();

    for (let num of [...nums1, ...nums2]) {
        if (maxHeap.size() === 0 || num <= maxHeap.peek()) {
            maxHeap.insert(num);
        } else {
            minHeap.insert(num);
        }

        if (maxHeap.size() > minHeap.size() + 1) {
            minHeap.insert(maxHeap.extractMax());
        } else if (minHeap.size() > maxHeap.size()) {
            maxHeap.insert(minHeap.extractMin());
        }
    }

    if (maxHeap.size() === minHeap.size()) {
        return (maxHeap.peek() + minHeap.peek()) / 2;
    } else {
        return maxHeap.peek();
    }
}

/**
 * Approach 5: Using Javascript Array Sort
 * Time Complexity: O((m+n)log(m+n))
 * Space Complexity: O(m+n)
 */
function findMedianSortedArrays_sort(nums1, nums2) {
    const mergedArray = [...nums1, ...nums2].sort((a, b) => a - b);
    const len = mergedArray.length;
    const mid = Math.floor(len / 2);

    if (len % 2 === 0) {
        return (mergedArray[mid - 1] + mergedArray[mid]) / 2;
    } else {
        return mergedArray[mid];
    }
}



// Example Usage and Verification
const nums1_1 = [1, 3];
const nums2_1 = [2];
console.log("Example 1 - Merge and Find:", findMedianSortedArrays_merge(nums1_1, nums2_1));
console.log("Example 1 - Optimized Merge:", findMedianSortedArrays_optimizedMerge(nums1_1, nums2_1));
console.log("Example 1 - Binary Search:", findMedianSortedArrays_binarySearch(nums1_1, nums2_1));
console.log("Example 1 - Heaps:", findMedianSortedArrays_heaps(nums1_1, nums2_1));
console.log("Example 1 - Sort:", findMedianSortedArrays_sort(nums1_1, nums2_1));


const nums1_2 = [1, 2];
const nums2_2 = [3, 4];
console.log("Example 2 - Merge and Find:", findMedianSortedArrays_merge(nums1_2, nums2_2));
console.log("Example 2 - Optimized Merge:", findMedianSortedArrays_optimizedMerge(nums1_2, nums2_2));
console.log("Example 2 - Binary Search:", findMedianSortedArrays_binarySearch(nums1_2, nums2_2));
console.log("Example 2 - Heaps:", findMedianSortedArrays_heaps(nums1_2, nums2_2));
console.log("Example 2 - Sort:", findMedianSortedArrays_sort(nums1_2, nums2_2));

const nums1_3 = [0, 0];
const nums2_3 = [0, 0];
console.log("Example 3 - Merge and Find:", findMedianSortedArrays_merge(nums1_3, nums2_3));
console.log("Example 3 - Optimized Merge:", findMedianSortedArrays_optimizedMerge(nums1_3, nums2_3));
console.log("Example 3 - Binary Search:", findMedianSortedArrays_binarySearch(nums1_3, nums2_3));
console.log("Example 3 - Heaps:", findMedianSortedArrays_heaps(nums1_3, nums2_3));
console.log("Example 3 - Sort:", findMedianSortedArrays_sort(nums1_3, nums2_3));

const nums1_4 = [2];
const nums2_4 = [];
console.log("Example 4 - Merge and Find:", findMedianSortedArrays_merge(nums1_4, nums2_4));
console.log("Example 4 - Optimized Merge:", findMedianSortedArrays_optimizedMerge(nums1_4, nums2_4));
console.log("Example 4 - Binary Search:", findMedianSortedArrays_binarySearch(nums1_4, nums2_4));
console.log("Example 4 - Heaps:", findMedianSortedArrays_heaps(nums1_4, nums2_4));
console.log("Example 4 - Sort:", findMedianSortedArrays_sort(nums1_4, nums2_4));

const nums1_5 = [3];
const nums2_5 = [-2, -1];
console.log("Example 5 - Merge and Find:", findMedianSortedArrays_merge(nums1_5, nums2_5));
console.log("Example 5 - Optimized Merge:", findMedianSortedArrays_optimizedMerge(nums1_5, nums2_5));
console.log("Example 5 - Binary Search:", findMedianSortedArrays_binarySearch(nums1_5, nums2_5));
console.log("Example 5 - Heaps:", findMedianSortedArrays_heaps(nums1_5, nums2_5));
console.log("Example 5 - Sort:", findMedianSortedArrays_sort(nums1_5, nums2_5));
